### ISSUE TITLE
Make transition to Engines subroute (in-repo) from the application throws error

### ISSUE THREAD

**Issue by sarathkumarbaskaran** [emberjs/ember.js#15809]
Let the engine mounted be 'test-engine' and 'list' be the route. Making transition to `test-engine/list` throws `"Assertion Failed: Attempting to inject an unknown injection: 'controller:application'".`

But if I make the transition to `test-engine` and `list` one by one. It works. Later the transition `test-engine/list` itself works.

Reproducible in > 2.14.

**Comment by pixelhandler** [comment]
@sarathkumarbaskaran can you make an example of the bug, perhaps in a github, gitlab or other shared git repository? One of the first steps to verifying the bug is to reproduce the behavior.

**Comment by sarathkumarbaskaran** [comment]
@pixelhandler Thanks for the note.

The issue happens only when I make reference to the Application's controller from the Engine Controllers and the application controller has an other controller injection. [Reproducible](https://github.com/sarathkumarbaskaran/sample-ember-app)

Example: 
Let `app/contollers/test.js` be

``` javascript
export default Controller.extend({
  appController: inject('application')
});
```

And `test-engine/addon/controllers/list.js`

``` javascript
import TestController from 'test-app/controllers/test';
export default TestController.extend();
```

Fails here `this.__container__.lookup(fullName, options);`
 `controller:application` is not registered in the container.

The use case is little weird here, just curious to know the change in behavior with the factoryFor lookup

**Comment by rwjblue** [comment]
**tl;dr** create `lib/engine-test/addon/controllers/application.js` as `export default Ember.Controller.extend()`.

---

This isn't really an issue with engines per-se, its that you cannot inject an auto-generated controller _before_ it has been generated. The reason it seems related to engines is that you are extending from an app controller _within_ the engine itself.

Here is a breakdown of what is going on when you transition into the nested `engine-test/list` route directly:

* router resolves engine-test engine
* router identifies target route (engine-test/list)
* router attempts to instantiate the route and controller for `engine-test/list`
  * note that `engine-test/list` controller is actually a reexport of the apps own `test` controller which includes `appController: inject('application')` 
  * when the controller is being instantiated, it checks to confirm that all of its required injections are available
  * it attempts to find `controller:application` inside the engine, but it is not found and therefore an error is thrown

When transitioning to `engine-test` _before_ transitioning to `engine-test/list` you force the `engine-test`'s `controller:application` to be created, which allows the injected controller check (when transitioning into `engine-test/list`) to succeed.

Closing for now, as I do not believe this is a bug here...

**Comment by sarathkumarbaskaran** [comment]
Thanks for the clarification 👍 . Just wanted to point out, it works with version <= ~2.13.

**Comment by sarathkumarbaskaran** [comment]
@rwjblue cool stuff 👍 . Just wanted to point out, it works with version <= ~2.13.
